The IPFS Project Roadmap (#1)
ipfs/roadmap: README.md

## IPFS Project Roadmap v0.4.0

## Table of Contents

- [IPFS Mission Statement](#ipfs-mission-statement)
- [2019 Priorities](#2019-priorities)
- [2019 Working Groups Roadmaps](#2019-working-groups-roadmaps)
- [2019 IPLD & libp2p Roadmaps](#2019-ipld--libp2p-roadmaps)
- [2019 Epics](#2019-epics)
- [2019 Goals (expanded)](#2019-goals-expanded)

>   - [📦 Package Managers](-package-managers-d1-e5-i3)

Reviewer: Missing #

## 2019 Priorities

We will be focusing our efforts into **5** top-level priorities:

- **📦 Supporting Package Managers**: Adding end-to-end support for package managers will provide a verifiable and co-hostable foundation for a vital use-case that we rely on. It will focus us on delivering performance and stability improvements while introducing IPFS to a huge community who have the skills to contribute back. See inbound interest [here](https://github.com/ipfs/notes/issues?q=is%3Aissue+is%3Aopen+label%3A%22package+managers%22).
  - _The most used code and binary Package Managers are powered by IPFS._
- **🗂 Transferring Large Files/Datasets**: Being the go-to solution for transferring large datasets is something IPFS is well poised to meet, but is currently blocked on performance, usability and pain-points that are also needed for future goals. Meeting this goal has the opportunity to embed IPFS in many tools with powerful ecosystems, or alternatively create improved versions that can attract new low-level developers to our community and create better solutions for end users.
  - _IPFS becomes the default way to distribute files / collections of files above 1GB._

> - **🔄 Building the Decentralized Web**: The decentralized web (DWeb) communities are already huge proponents and users of IPFS, however they need additional tooling and capabilities to spark rapid growth of DWeb Applications. Augmenting the power and number of DWeb Applications created will increase the size and impact of our existing community, and test IPFS in ways only our users can imagine.

Reviewer 2: ```suggestion
- **🔄 Building the Decentralized Web**: IPFS wants to decentralize the web. The communities building the DWeb are already huge proponents of IPFS. By providing additional tools and capabilities we can support and amplify the movement. Augmenting the power and number of DWeb Applications created will increase the size and impact of our existing community, and test IPFS in ways only our users can imagine.
```

  - _IPFS supports decentralized web apps built on p2p connections with power and capabilities at the edges._
- **🤝 Supporting community growth**: We will ensure that we are setting up our community for success by addressing key needs to unlock their growth in 2019. Making our users successful and delivering on their top needs to ensure they have a smooth and productive experience with IPFS helps us retain and gain adoption, grow our community, increase our impact, and demonstrate our maturity as a project.
  - _IPFS users are set up with the tools and capabilities they need to grow & flourish._
- **🧠 Building the foundation for future goals**: In addition to our focused 2019 project goals, we have many [future goals](#2020-goals) for IPFS that we want to build toward. Some of the work required for those later goals requires advance planning to set up time-sensitive projects or partnerships for the future. This priority encompasses the forward-looking work we strategically take on now to set ourselves up for success later.
  - _IPFS takes on strategic projects to stay on-track to meet future goals._

> **Why these priorities**

Reviewer: Add "?" at the end?

The 2019 priorities are all geared to grow our maturity as a project, increase our adoption and community, and set ourselves up for the future. To be successful, we need to level-up our ability to land complete, polished, and accessible improvements that meet our user's needs for stability, scalability, privacy, and security. We need to keep growing intelligently, both by the types of communities that can help us accelerate growth and by the number of users they can help us reach. And we need to set ourselves up with the understanding and capabilities to reach our future goals for the project and scale our impact.

> We designed a sorting function to help us identify which goals 1) we could focus on landing end-to-end in 2019, 2) that would bring valuable new partners and users into our community, and 3) that would set us up for future goals while delivering near-term value.

Reviewer 2: ```suggestion
We designed a sorting function to help us identify goals that 1) we could focus on landing end-to-end in 2019, 2) would bring new partners and users into the community, and 3) would set us up for future goals while delivering near-term value.
```

> We designed a [sorting function](#heading=h.cr92wqlur7kk) to help us identify which goals 1) we could focus on landing end-to-end in 2019, 2) that would bring valuable new partners and users into our community, and 3) that would set us up for future goals while delivering near-term value.

Reviewer: Link not working

**Sorting Function Explanation**

> **D** = Difficulty (or "Delta" or "Distance"), **E** = Ecosystem Growth, **I** = Importance

Each goal was given a score from 1 (low) - 5 (high) on each axis. We sorted first in terms of low difficulty or "delta" (i.e.  minimal additional requirements and fewer dependencies from the capabilities IPFS has now), then high ecosystem growth (growing our community and resources to help us gravity assist and accelerate our progress), and finally high importance (we want IPFS to have a strong, positive impact on the world). See the [2019 Goals (expanded)](#2019-goals-expanded) and [2020+ Goals](#2020-goals) sections for sorted priorities.

> ## 2019 Working Groups Roadmaps

Reviewer 2: This list would benefit from an introduction. Right now the doc goes from high-level goals to a list of working group links, without explaining the link.

Author: Good point! @olizilla would you like to volunteer and submit a PR to add that introduction?

- [Project WG](<link when the file is on gh>)
- [JS Core](<link when the file is on gh>)
- [Go Core](<link when the file is on gh>)
- [GUI](<link when the file is on gh>)
- [Cluster](<link when the file is on gh>)
- [IPFS Infrastructure](<link when the file is on gh>)
- [Community](<link when the file is on gh>)
- [Integration in Web Browsers](<link when the file is on gh>)
- [Dynamic Data & Capabilities](<link when the file is on gh>)
- [Decentralized Data Stewardship](<link when the file is on gh>)

## 2019 IPLD & libp2p Roadmaps

- [libp2p](https://docs.google.com/document/d/1Rd4yNw1TNQBvfRrKeEMSTseb6fvPzS-C--obOn0nul8/edit)
- [IPLD](https://github.com/ipld/roadmap/pull/1)

## 2019 Epics

> We've distilled the Working Group Milestones into 2019 IPFS Epics in order to create a overview of the primary targets the IPFS Project is setting for 2019. The information about Timeline, Sequencing and Priority still resides within each Working Group Roadmap.

Reviewer 2: ```suggestion
We've distilled the key themes from the Working Group Roadmaps into a list Epic Endevours that give an overview of the primary targets IPFS has for 2019. The details about the tasks and priorities that get us there are defined in the Working Group Roadmaps.
```

> We've coalesced the Working Group Milestones into 2019 IPFS Epics in order to create a overview of the primary targets the IPFS Project is setting for 2019. The information about Timeline, Sequencing and Priority still resides within each Working Group Roadmap.

Reviewer: Suggestion: coalesced -> distilled. This isn't just all of them bundled together as one, these epics have extracted the essential meaning and most important aspects of the WG milestones.

If you are pumped about these Epics and want to help, you can also get involved! See the call to action (CTA) for each section below.

> **1. The reference implementations of the IPFS Protocol (Go & JS) become Production Ready**

Reviewer: become -> becomes

Reviewer: Suggestion: make these headings so we can link to them.

> We've been working hard on improving and iterating the IPFS Protocol to make sure it serves its users needs and achieves the goal of giving the Web a new super power, Content Addressing. In 2019, we are setting ourselves to solidify the core components and ensure IPFS is ready for Production Systems. Some of these goals include:

Reviewer 2: ```suggestion
We've been working hard on improving and iterating on the IPFS Protocol to make sure it serves its users needs and achieves the goal of giving the Web a new super power: Content Addressing. In 2019 we will solidify the core components and ensure IPFS is ready for production systems. Our goals include:
```

- IPFS core solidified and APIs revamped/crystalized to be future-proof
- Complete Specification of the Protocol
- Security Audits and Security Program

> - Go and JS IPFS enable modern IPFS data formats (UnixFSv2, CIDv1, raw blocks) by default

Reviewer 2: Link to specs here...
```suggestion
- Go and JS IPFS enable modern IPFS data formats ([UnixFSv2](https://github.com/ipfs/unixfs-v2), [CIDv1](https://github.com/ipld/cid), raw blocks) by default
```

- Bring js-ipfs and go-ipfs to v1.0.0

> You can contribute to these Epic by testing, debugging or even coding to either Go or JS IPFS. Both teams have a chat every week that you can join. [Go](https://github.com/ipfs/team-mgmt/blob/master/MGMT_GOLANG_CORE_DEV.md) & [JS](https://github.com/ipfs/team-mgmt/blob/master/MGMT_JS_CORE_DEV.md#weekly-core-dev-team-calls).

Reviewer 2: ```suggestion
You can contribute to these Epic by testing, debugging, documenting or coding on either Go or JS IPFS. Both teams have a chat every week that you can join. [Go](https://github.com/ipfs/team-mgmt/blob/master/MGMT_GOLANG_CORE_DEV.md) & [JS](https://github.com/ipfs/team-mgmt/blob/master/MGMT_JS_CORE_DEV.md#weekly-core-dev-team-calls).
```

> You can contribute to these Epic by testing, debugging or even coding to either Go or JS IPFS. Both teams have a chat every week that you can join. [Go](https://github.com/ipfs/team-mgmt/blob/master/MGMT_GOLANG_CORE_DEV.md) & [JS](https://github.com/ipfs/team-mgmt/blob/master/MGMT_JS_CORE_DEV.md#weekly-core-dev-team-calls).

Reviewer: "Both teams have a chat" => "Both teams have a meet up"?

**2. Support Software Package Managers into entering the Distributed Web**

> Software Package Managers are one of the best systems to take advantage of Content Addressing. Developers and Deployment systems hit Package Managers Registries everyday to download the same bits of code, creating a huge overhead in terms of bandwidth. In addition today, it is incredibly important that we ensure that the code that we run is the code that the author's intended and Content Addressing can help with this goal by adding a Integrity verification step.

Reviewer: "Package Managers Registries" => "Package Manager's registries"

> - Kick-off a Package Managers Working Group to work closely with different Package Managers and understand each one challenges and needs.

Reviewer: "understand each one challenges" => "understand each ones challenges"

- Publish a report on the results from adding the NPM Registry to IPFS.

> - Meet the scalability and performance needs to serve the Package Managers users.

Reviewer: "Package Managers users" => "Package Manager's users"

> We will booting the Package Managers Working Group soon, meanwhile, check out [NPM on IPFS](https://github.com/ipfs-shipyard/npm-on-ipfs) & [GX](https://github.com/whyrusleeping/gx) to get a feel for what IPFS can contribute to Package Management.

Reviewer: "We will booting the" => "We will be booting the"

Reviewer: "soon" => "in early 2019"?

**3. Scale the IPFS Network**

To support the continued growth of the IPFS Network, we plan to optimize our scalability and testing.
- Deploy a P2P testbed for Distributed Protocols that enables us to spin up hundreds of thousands of nodes on demand and run tests on them.

> - IPFS has an opt-in Telemetry system so that tests can collect key data to understand the experiments.

Reviewer: hmm, I always figured this was more about us understanding how IPFS is growing. Opt-in sounds disabled by default (not what we want), but I think opt-out is not right as I'd like to think we'd confirm with people before just collecting and reporting them. Maybe "optional" is a better word?

Our friends from libp2p will be playing the larger role into achieving this goal. If you are a Networks specialist and/or if you like scaling systems to serve hundreds of millions, join the conversation at [Bi-Weekly Call](https://github.com/libp2p/team-mgmt#bi-weekly-call) and learn more.

> **4. Create a platform that untangles the mysteries of developing a DApp**

Reviewer: Needs an explaination of what "DApp" stands for.

Support the explosion of the DApp ecosystem with underlying tooling, best practices, and community support.

> - Fully launch Peer Star (Peer Base) toolkit to support building Dapps that can run Online, Offline, and in Disjoint Networks.

Reviewer: "Dapps" => "DApps"

> - Translate the Identity Manager RFC into code, a sandboxed way for Users to manage their DWeb Identity.

Reviewer: "Users" => "users"?

- Support DApp security needs (API security, multi-tenancy, and secure broadcast updates).
- Create a guidebook/spec to popularize best practices to deal with unique UX challenges of the DWeb.

> - Continue to contribute to the adoption of the DWeb in the WebPlatform (Brave, libdweb, IPFS companion), making it more accessible to users and Web Developers.

Reviewer: Is "WebPlatform" meant to be one word?

...I didn't realise that was a thing...I have no idea what I'm doing dog <--me

> The Dynamic Data & Capabilities WG is working full speed towards this. If you are passionate about building DApps and/or have field experience you would like to share, participate the [DDC Bi-weekly Sync.](https://github.com/ipfs/dynamic-data-and-capabilities#bi-weekly-sync)

Reviewer: "participate the" => "please come and participate in the"

- The Migration to Base32 CIDs is completed across all projects and IPFS Gateway.
- Cube, the DIY IPFS pinning service, is built and released.
- WebUI shows information about file transfer progress, files being stored on our node (independently of that list size) and enables users to share files with friends and/or local peers.
- Managed and federated clusters become easy to boot and administer.
- Support self-hosted topic networks with tooling for collaborative/reciprocal pinsets.
- Create a UI for LAN/WAN Sharing on IPFS Companion and IPFS Desktop.

**7. IPFS testing, benchmarks, and performance optimizations**

In order to prepare ourselves for production readiness and ensure IPFS is optimized for a variety of real-world use cases, we'll double down on testing and benchmarking in 2019.
- Testing harness and environments that simulate different topologies / network configurations.
- Tests for Interoperability, Performance Benchmarks, Reliability, and Network churn.
- Performance overhaul for resolving (IPNS), finding (Providers), transferring, and adding files.

**8. Support the growing IPFS Community**

> The IPFS Community is composed of many enthusiasts from the DWeb, P2P, Crypto, Privacy, builders, Web 3.0, Blockchain and many more. Today we have an impressive 3000++ contributors directly improving the IPFS, libp2p, IPLD and Multiformats repos, but there are tens of thousands more contributing to Apps, Libraries and Systems that build on top of IPFS everyday! In 2019, we want to provide a rich platform for all these contributors to share their ideas, get their questions heard, cross-pollinate and tackle new challenges!

Reviewer: "3000++" => "3,000+"?

- The documentation on IPFS and its libraries gets revamped and becomes part of continuous delivery.
- Increase awareness and adoption through Community outreach via presentations, guides / how-tos, meetups, and quarterly IPFS Project updates.
- Support contributor productivity and impact through Quarterly User Reports, fast code reviews, and DevOps automation.
- Create IPFS Dev Grant and Research Grant programs to encourage global engagement with the cutting-edge work in the IPFS ecosystem.

> - Launch ProtoSchool, an interactive way to learn how to build on the P2P Web.

Reviewer: Link to the org?

### 💾 Packet Switched Web (D3 E2 I3)

> IPFS protocols use packet switching, and the network can relay all kinds of traffic easily, tolerating switch failures.

The infrastructure protocols (libp2p, IPFS, etc.) and the end-user app protocols (the logic of the app) can work entirely over a packet switching layer. Protocols like BitSwap, DHT, PubSub become drastically higher performance, and unconstrained by packets sent before theirs. Web applications can form their own isolated virtual networks, allowing their users to distribute the packets. Users can form their own groups and their own virtual networks, allowing users to only operate in a subnet they trust, and ensure all of their traffic is moving between trusted switches. The big public network uses packet switching by default.

### 📑 Data Web (D4 E3 I3)

> Large Datasets are open, easy to access, easy to replicate, version controlled, secure, permanent.

> `TODO: Write full description`

Reviewer 3: Before we merge either add the description or remove this line :)
```suggestion

```
